internal/injector/parse.go

```diff
@@ -0,0 +1,137 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023-present Datadog, Inc.
+
+package injector
+
+import (
+	"bytes"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"io"
+	"os"
+)
+
+// parseFiles parses all provided files, after having applied any leading
+// "//line" directives if present.
+func (i *Injector) parseFiles(fset *token.FileSet, files []string) ([]*ast.File, error) {
+	result := make([]*ast.File, len(files))
+	for idx, file := range files {
+		var err error
+		result[idx], err = i.parseFile(fset, file)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return result, nil
+}
+
+// parseFile parses the provided filename, after having applied a leading
+// "//line" directive if one is present.
+func (i *Injector) parseFile(fset *token.FileSet, filename string) (*ast.File, error) {
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, fmt.Errorf("open %q: %w", filename, err)
+	}
+	defer file.Close()
+
+	// If the file begins with a "//line <path>:1:1" directive, we consume it and
+	// then pretend the "<path>" was our filename all along. This simplifies
+	// handling of line offsets further down the line and removes some duplicated
+	// effort to do it early.
+	mappedFilename := filename
+	if mapped, err := consumeLineDirective(file); err != nil {
+		return nil, fmt.Errorf("peeking at first line of %q: %w", filename, err)
+	} else if mapped != "" {
+		mappedFilename = mapped
+	}
+
+	astFile, err := parser.ParseFile(fset, mappedFilename, file, parser.ParseComments)
+	if err != nil {
+		return nil, fmt.Errorf("parsing %q: %w", filename, err)
+	}
+
+	return astFile, nil
+}
+
+// consumeLineDirective consumes the first line from r if it's a "//line"
+// directive that either does not have line/column information or has it set to
+// line 1 (and column 1). If the directive is consumed, the filename it refers
+// to is returned. Otherwise, the reader is rewinded to its original position.
+func consumeLineDirective(r io.ReadSeeker) (string, error) {
+	var buf [7]byte
+	n, err := r.Read(buf[:])
+	if err != nil {
+		return "", err
+	}
+	if string(buf[:n]) != "//line " {
+		_, err := r.Seek(0, io.SeekStart)
+		return "", err
+	}
+
+	buffer := make([]byte, 0, 128)
+	var wasCR, done bool
+	for !done {
+		if n, err := r.Read(buf[:1]); err != nil {
+			return "", err
+		} else if n == 0 {
+			// Reached EOF
+			break
+		}
+		switch c := buf[0]; c {
+		case '\n':
+			done = true
+		case '\r':
+			wasCR = true
+			continue
+		default:
+			if wasCR {
+				// We saw a CR and this is not an LF, so we rewind one byte and bail out.
+				if _, err := r.Seek(-1, io.SeekCurrent); err != nil {
+					return "", err
+				}
+				done = true
+			} else {
+				buffer = append(buffer, c)
+			}
+		}
+	}
```

Comment:
I am trying to understand what is going on here. Why not use `bufio.Reader.ReadLine` instead doing this? We don't really care if the CRLF is broken do we? After we read the directive we can just seek to the end of the line no? :thinking:

Reply:
Basically, `bufio.Reader.ReadLine` will read more than just the line (it's buffered), so if we have a match, then we need to seek back to the beginning of the file, then forward again past the directive and also the EOL sequence after that...

It's not a question of CRLF being broken, it's about there being 3 possible EOL sequences (CR, LF, CR+LF), so if we see a CR, we only consume the next char if it's an LF.

This approach seemed more efficient overall than going with buffered IO, especially in that it involves fewer seeks & has completely deterministic file offsets at each point.